pkg/server/autoconfig/auto_config.go

```diff
@@ -40,33 +53,216 @@ func (r *autoConfigRunner) Resume(ctx context.Context, execCtx interface{}) erro
+	var taskQueue []autoconfigpb.Task
 
-	case <-shutdownCh:
-		return context.Canceled
+	for {
+		if len(taskQueue) == 0 {
+			// No tasks to create. Just wait until some tasks are delivered.
+			log.Infof(ctx, "waiting for more tasks...")
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-shutdownCh:
+				return context.Canceled
 
-	case <-wait:
+			case <-waitForMoreTasks:
+				taskQueue = provider.GetTasks()
+				continue
+			}
+		}
+
+		r.job.MarkIdle(false)
+		var waitSome bool
+		var err error
+		waitSome, taskQueue, err = maybeRunNextTask(ctx, provider, execCfg, taskQueue)
+		if err != nil {
+			log.Warningf(ctx, "error processing auto config: %v", err)
+		}
+		r.job.MarkIdle(true)
+		if err != nil || waitSome {
+			// Wait a bit before retrying.
+			select {
+			case <-time.After(5 * time.Second):
+				continue
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-shutdownCh:
+				return context.Canceled
+			}
+		}
 	}
+}
+
+// maybeRunNextTask checks which task was last executed, ratchets the
+// task queue beyond the last completed tasks, and attempts to start
+// the first next task in the queue.
+func maybeRunNextTask(
+	ctx context.Context,
+	provider acprovider.Provider,
+	execCfg *sql.ExecutorConfig,
+	taskQueue []autoconfigpb.Task,
+) (waitSome bool, remainingTaskQueue []autoconfigpb.Task, err error) {
+	log.Infof(ctx, "got %d tasks to create", len(taskQueue))
+
+	// First of all, wait on any task jobs, if any.
+	//
+	// This is an optimization: the logic below would work without this step.
+	// We do this to avoid waiting too much after a task has completed
+	// (the logic below waits for a random amount of time in case of conflict).
+	if err = maybeWaitForCurrentTaskJob(ctx, execCfg); err != nil {
+		return true, taskQueue, err
+	}
+
+	// Now on to the actual work.
+	//
+	// Transactionally, we are going to extract the last known
+	// completed task; ratchet our queue forward to that point; assert
+	// there is no already-started task beyond it; then create our
+	// job.
+
+	// The job ID we'll use.
+	jobID := execCfg.JobRegistry.MakeJobID()
+	log.Infof(ctx, "allocated job ID %d for next task (of %d)", jobID, len(taskQueue))
+	var nextTaskID uint64
+
+	err = execCfg.InternalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) (resErr error) {
+		// Is there any other started task already?
+		otherTaskID, _, err := getCurrentlyStartedTaskID(ctx, txn, execCfg)
+		if err != nil {
+			return err
+		}
+		if otherTaskID != 0 {
+			// What likely happened is that another node caught up with us,
+			// observed the job completion for the previous task and already
+			// created the start marker and job for the next one.
+			log.Infof(ctx, "found start marker for task %d, unable to start another task", otherTaskID)
+			// We will simply retry, with some jittered delay.
+			waitSome = true
+			return nil
+		}
+
+		// Find the latest completed task.
+		lastTaskID, err := getLastCompletedTaskID(ctx, txn, execCfg)
```

Comment:
It sort of feels like it'd be nice to combine the fetching of the currently started task with the fetching of the latest completed task.